### ISSUE TITLE
a length check fails (_axes.py @ 4386): new to 3.2.0; not present in 3.1.3

### ISSUE THREAD

**Issue by e-gun** [matplotlib/matplotlib#16739]
### Bug report

ValeuError exception when sending array to matplotlib via networkx: 

code that generated arrays that used to pass through matplotlib without a problem now yields exceptions

a length check now fails (_axes.py @ 4386):

```
        s = np.ma.ravel(s)
        if len(s) not in (1, x.size):
            raise ValueError("s must be a scalar, or the same size as x and y")
```

**Code for reproduction**

`networx` is the required dependency
```
import networkx as nx
```

the following will trigger the exception:
```
scalednodes = [6740.484952926636, 4175.739884376526, 4102.685809135437, 4071.070432662964, 4070.39737701416, 4004.2176246643066, 3979.384660720825, 3937.743902206421, 3920.8098649978638, 3871.6392517089844, 3852.8380393981934, 3833.83047580719, 3821.3353157043457, 3817.2022104263306, 3790.743112564087]

nx.draw_networkx_nodes(graph, nx.fruchterman_reingold_layout(graph), node_size=scalednodes, alpha=0.75, node_color=range(len(terms)), cmap='Pastel1')
```

**Actual outcome**

```
  File "[mypath]/semanticvectors/vectorgraphing.py", line 162, in matplotgraphmatches
    nx.draw_networkx_nodes(graph, pos, node_size=scalednodes, alpha=0.75, node_color=range(len(terms)), cmap='Pastel1')
  File "[myvenv]/lib/python3.7/site-packages/networkx/drawing/nx_pylab.py", line 417, in draw_networkx_nodes
    label=label)
  File "[myvenv]/lib/python3.7/site-packages/matplotlib/__init__.py", line 1543, in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
  File "[myvenv]/lib/python3.7/site-packages/matplotlib/cbook/deprecation.py", line 358, in wrapper
    return func(*args, **kwargs)
  File "[myvenv]/lib/python3.7/site-packages/matplotlib/axes/_axes.py", line 4387, in scatter
    raise ValueError("s must be a scalar, or the same size as x and y")
```

```
# If applicable, paste the console output here
#
#
```

**Expected outcome**

**Matplotlib version**

  * Operating system: `macOS` & `linux`
  * Matplotlib version: `3.2.0` but not `3.1.1`
  * Matplotlib backend (`print(matplotlib.get_backend())`): `MacOSX`
  * Python version: `3.7.3`
  * Jupyter version (if applicable):
  * Other libraries: `networx`

installed via `pip`

**Comment by e-gun** [comment]
PS: we are off by one when we do the check:

`_axes.py 4386: len(s), x.size 15 16`

**Comment by QuLogic** [comment]
Please provide a _complete_ example that reproduces the bug, preferably without networkx, but at the very least without external data.

**Comment by e-gun** [comment]
**MATPLOTLIB 3.1.3**

```
from random import uniform as rnd
import matplotlib.pyplot as plt
ax = plt.gca()

a = [-0.10902419, -0.75646717, -0.96403103, -0.1960284,  -0.05789045,  0.16462492, 0.25156931, 0.56237587, 0.69960643, 0.31271061, -0.62216608, 0.47206789, 0.60326695, -0.08107401, -0.45185505, 0.17231438]

b = [-0.07969323, -0.32750072, -0.09406059, 0.27041506, -1., 0.18910032, 0.62951915, 0.09861711, -0.06275149, 0.22269021, -0.86167041, -0.23924364, 0.42907668, 0.58223968, 0.54558958, -0.30232771]

c = [9323.340654373169, 6616.982460021973, 5690.618991851807, 5689.395904541016, 5672.798752784729, 5638.451099395752, 5606.265664100647, 5553.000926971436, 5544.699668884277, 5531.859755516052, 5476.373791694641, 5463.115096092224, 5454.218745231628, 5441.088080406189, 5435.509443283081]

>>> ax.scatter(a, b, c)
<matplotlib.collections.PathCollection object at 0x116fe1ef0>

d = [rnd(-1,1) for _ in range(16)]
e = [rnd(-1,1) for _ in range(16)]
f = [rnd(3000, 4000) for _ in range(15)]

>>> ax.scatter(d, e, f)
<matplotlib.collections.PathCollection object at 0x11e340518>


```
**MATPLOTLIB 3.2.0:**

[take the same data/code]

```
>>> ax.scatter(d, e, f)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/erik/hipparchia_venv/lib/python3.7/site-packages/matplotlib/__init__.py", line 1543, in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
  File "/Users/erik/hipparchia_venv/lib/python3.7/site-packages/matplotlib/cbook/deprecation.py", line 358, in wrapper
    return func(*args, **kwargs)
  File "/Users/erik/hipparchia_venv/lib/python3.7/site-packages/matplotlib/axes/_axes.py", line 4387, in scatter
    raise ValueError("s must be a scalar, or the same size as x and y")

```

**Comment by QuLogic** [comment]
You have one less element in `f` than in `d` and `e`; what do you expect it to do?

**Comment by e-gun** [comment]
Well, I suppose I do mean for `f` to have more than one element thatn `d` and `e`. Specifically, I have been graphing via `networx` and have not asked how and/or why. I just send it my data. Then it calls `matplotlib`. Then I get the graph. 

This has been going on for a couple of years now. But then things broke between `3.1.3` and `3.2.0`. I submitted a bug report. You told me not to mention `networx`. So I followed my data down through `networx`, found the spot where it `networx` calls `matplotlib` and then generated a self-standing reproducible bug. 

As you can see, if you send the same data to `matplotlib 3.1.3` and to `matplotlib 3.2.0` you get different results.

I can definitely see how the source code for `_axes.py` is indeed checking for a length match. What I do not understand is why the exception is new. I have little/no appreciation of the inner workings of either `networx` or `matplotlib`, but I do know that I used to get graphs with this sort of data and now I do not...

In short, there is not much more I can do to help you. Sorry about that. But perhaps this code will at least let you decide what the underlying issue is.

**Comment by e-gun** [comment]
PS: If I replace line 4387 I can have my graphs...

old: 
`raise ValueError("s must be a scalar, or the same size as x and y")`

temp fix:
`print('not raising ValueError...')`

**Comment by QuLogic** [comment]
The check is new because it was added to catch bugs like this one. Previous versions produced graphs that were, in almost all cases, non-sensical. The bug here is in networkx, as it is passing incorrect values to Matplotlib.

**Comment by e-gun** [comment]
You closed the issue, but the graphs were not non-sensical. They made perfect sense. Now they do not get drawn at all.

The graph had 16 dots. One was in the center. 15 of them were connected to the center dot. 

So I can see how an array of 16 and an array of 15 could/should co-exist in such a situation. 

I will contact networx about this. But I am not exactly convinced that their code is “buggy” given that it was producing good/expected graphs…

**Comment by ImportanceOfBeingErnest** [comment]
To me it appears that this is due to an API change, which did not undergo deprecation as usual. While one could argue that providing 16 points and 15 sizes is nonsensical, the API actually did allow it, and there are a couple of cases where it makes much more sense, e.g. think of `len(s)` being a divisor of `len(x)`.   
Reopening this to get one or two more opinions on whether the error should be turned into a deprecation warning for 3.2.1.

**Comment by tacaswell** [comment]
Apparently we used to just wrap the sizes ?!

```python
c = (np.arange(15) + 1) * 5000                                                                                                                              
a = b = np.arange(16)                                                                                                                                       
plt.scatter(a, b, c, alpha=.2)  
```
![Figure_1](https://user-images.githubusercontent.com/199813/76671368-511dc400-656c-11ea-876b-2cb37a5de0d9.png)

If the lengths are multiples of each other, but off-by-one seems like a bug on the side of the caller.

@e-gun I took a quick look at networkx's github and could not find an issue from you there.

**Comment by tacaswell** [comment]
On a bit more investigation, I don't think this is a bug in networkx, the `node_size` is passed in by the user and their docs say

```python
node_size : scalar or array
   Size of nodes (default=300).  If an array is specified it must be the
   same length as nodelist.
```

@e-gun You should pad your scaled_node list by 1 and then it should work with mpl 3.2

**Comment by stonebig** [comment]
just as a remark, the previous plot is obtained from :
````
c = (np.arange(16) + 1) * 500                                                                                                                              
a = b = np.arange(16)                                                                                                                                       
plt.scatter(a, b, c, alpha=.2)  
````

**Comment by e-gun** [comment]
@tacaswell: thank you very much! 

This is a very simple solution to execute on my end: `scalednodes.append(0)`

 And it seems to yield the sorts of graphs I expect to see. 

It also works with `matplotlib 3.1.3` and `matplotlib 3.2.0` so I don't have to do version tracking.

![graph-2](https://user-images.githubusercontent.com/20753804/76687103-317fad80-65f7-11ea-958b-777087bade95.png)

**Comment by jklymak** [comment]
Closing because this appears to have been something that just happened to work by accident and not really a matplotlib bug per-se.  @e-gun, you may still want to open something with network-x that they may want to check arguments more strictly.
